### ISSUE TITLE
Acceptance rate on 2.0 is very low

### ISSUE THREAD

**Issue by ABLomas** [Anbeeld/beellama.cpp#26]
Compiled 2.0 release with Cuda compilation tools, release 13.2, V13.2.78 Build cuda_13.2.r13.2/compiler.37668154_0
Tried exactly as docs says (because using single RTX3090 24GB) on ubuntu revolute:

 -  Qwen3.6-27B-Q5_K_S.gguf
 - Qwen3.6-27B-DFlash-Q4_K_M.gguf
 - mmproj-BF16.gguf (but this one does not matter much)

Full cmdline (with some tweaks already, initially tried suggested temp):
```
beellama.cpp/build/bin/llama-server \
  -m "/home/models/llm/beelama/Qwen3.6-27B-Q5_K_S.gguf" \
  --mmproj "/home/models/llm/beelama/mmproj-BF16.gguf" \
  --no-mmproj-offload \
  --spec-draft-model "/home/models/llm/beelama/Qwen3.6-27B-DFlash-Q4_K_M.gguf" \
  --spec-type dflash \
  --spec-dflash-cross-ctx 1024 \
  --port 8082 \
  --host 0.0.0.0 \
  -np 1 \
  --kv-unified \
  -ngl all \
  --spec-draft-ngl all \
  -b 2048 -ub 512 \
  --ctx-size 134000 \
  --cache-type-k q5_0 --cache-type-v q4_1 \
  --flash-attn on \
  --cache-ram 0 \
  --jinja \
  --no-mmap --mlock \
  --no-host \
  --reasoning on \
  --chat-template-kwargs '{"preserve_thinking":true}' \
  --temp 0.4 --top-k 20 --top-p 1.0 --min-p 0.0

```
What i get - very low acceptance rate (if i understand statistics correctly) - lower than 20% in all cases.
Then i tried to reduce temp to zero, increase top-k, still:

```
slot launch_slot_: id  0 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> ?top-p -> ?min-p -> ?xtc -> temp-ext -> dist
slot launch_slot_: id  0 | task -1 | spec dm controller: adaptive=1 controller=profit probe_fraction=0.2500 explore_interval=12 profit_min=0.0500 raise=0.0500 lower=0.0500 ewma=0.1500 min_samples=3 warmup=0 baseline_interval=1024 p_min=0.0000 draft_p_min=0.0000
slot launch_slot_: id  0 | task -1 | request sampling: n_predict=-1 ignore_eos=0 stop=0 reasoning_budget=-1 temp=0.000 dynatemp_range=0.000 dynatemp_exponent=1.000 top_k=1000 top_p=1.000 min_p=0.000 typ_p=1.000 top_n_sigma=-1.000 mirostat=0 samplers=penalties;dry;top_n_sigma;top_k;typ_p;top_p;min_p;xtc;temperature reasoning_loop_guard=1 loop_min=1024 loop_window=2048 loop_period=512 loop_coverage=768 loop_interval=32 loop_interventions=1 repeat_penalty=1.000 presence_penalty=0.000 frequency_penalty=0.000 dry_multiplier=0.000 dry_allowed_length=2 adaptive_n_max=14
<...>
slot   operator(): id  0 | task 9724 | adaptive dm profit: cur=0 recommended=14 score=24.9 action=apply
reasoning-budget: deactivated (natural end)
slot   operator(): id  0 | task 9724 | adaptive dm profit: cur=14 recommended=12 score=44.6 action=apply
slot print_timing: id  0 | task 9724 |
prompt eval time =    1449.04 ms /   910 tokens (    1.59 ms per token,   628.00 tokens per second)
       eval time =   79974.83 ms /  3146 tokens (   25.42 ms per token,    39.34 tokens per second)
      total time =   81423.87 ms /  4056 tokens
draft acceptance rate = 0.14337 ( 2088 accepted / 14564 generated)
adaptive dm: fringe=0.00 n_max=12
statistics dflash: #calls(b,g,a) = 12 10703 8043, #gen drafts = 10703, #acc drafts = 8043, #gen tokens = 128319, #acc tokens = 22314, dur(b,g,a) = 0.014, 93212.216, 0.471 ms

```
Speed is not bad (i have 250W power cap) but i kinda expect better acceptance rate.
I tried different prompts - from "daily stuff" to "fix this yaml" or "write script..." - acceptance rate varies slightly, but usually between 12 and 18%.
ToDo (to test) - increase top-k to 10K or so, increase --spec-dflash-cross-ctx - but better to know if settings should be changed between versions (and provided defaults are no longer correct)?

**Comment by Anbeeld** [comment]
The reason it has "low" acceptance is because it drafts a lot of tokens. Default draft-n-max is 16 and it's adaptive, meanwhile for mainline llama.cpp MTP it's 3 fixed. Drafting was made cheap back in v0.1.2 so it's more profitable to try more, even if most fail.

Check benchmarks in the readme, for example. The actual important number is the second one in the acceptable column, *accepted draft tokens to final generated tokens*.

So we draft a lot of tokens, and maybe only a few of them are accepted, but the ones that are accepted are quite useful. It's just that the input acceptance looks "bad" compared to MTP because it tries much longer chains.

So in the end one should not blindly compare acceptance between different systems. The actual tok/s is a better metric in this cases. Raw acceptance might be useful to compare within same system, e.g. when tinkering with config as you suggest and as I encourage you to do so, but even then tok/s is the main metric and acceptance is more of a secondary diagnostic that also must be adjusted to how drafting works.

And then there's the "it depends on the prompt", because it really does.

**Comment by ABLomas** [comment]
OK, maybe title is misleading - should be "lower speeds compared to benchmark table".
I used example (but probably benchmarks should not be run with example settings, as temperature is obviously too high (0.7 as in example) - lots of loops and mistakes):
| **Task** | **Acceptance Rate** | **Prompt Speed** | **Output Speed** | 
| :--- | :--- | :--- | :--- |
| **Task store** | **53.5%** | 825.5 tok/s | **115.6 tok/s** | 
| **KV report** | **55.0%** | 1243.4 tok/s | **105.4 tok/s** |
| **Doubly-linked** | **48.6%** | 446.7 tok/s | **91.8 tok/s** |
| **Multi-turn** | **10.4%** | 1001.2 tok/s | **38.8 tok/s** |

So in short (i set stock power limits on card):
Task Store: i got 115.6 tok/s, you 163.9 tok/s
KV Report: i got 105.4 tok/s, you 157.7 tok/s

The logs show reasoning-loop-guard: triggered multiple times but this is due to temperature most likely.
Multi-turn - logs show adaptive dm: disabling speculative depth repeatedly so it was ran on baseline speeds

Then i tried temp=0.1
I expected looping and errors to be gone, but no - they are still here, from quick look none of programs written are ok (even basic syntax). Strings in hebrew, chinese, looping on the same word - all that carry from temp=0.7 and is still here, but performance dropped even more:

| **Task** | **Acceptance Rate** | **Prompt Speed** | **Output Speed** | 
| :--- | :--- | :--- | :--- |
| **Task store** | **48.3%** | 773.0 tok/s | **79.4 tok/s** | 
| **KV report** | **35.2%** | 707.0 tok/s | **60.6 tok/s** | 
| **Doubly-linked** | **42.8%** | 384.2 tok/s | **78.2 tok/s** | 
| **Multi-turn** | **30.7%** | 836.8 tok/s | **52.6 tok/s** | 

So this is way lower than in benchmark table (and still contain lots of errors)

Error examples at temp=0.1
```
       def test_find(self):
        """Test the find method."""
        dll = DoublyLinkedList[int]()
        dll.append(1)
        dll.append(2)
        dll.append(3)
        assert dll.find(2) == 1بور
بور       بور assertبور dllبور.بورfindبور(4بور)بور ==بور -1
    def test_reverse(self):
<...>
    def render_lines(self) -> List[str]:
        result: List[str] = [] for row in self._rowsEndian: eliminated
 eliminated            result eliminated.append eliminated(f eliminated"{ eliminatedrow eliminated.key eliminated}: eliminated{ eliminatedrow eliminated.value eliminated}" eliminated) eliminated
 eliminated        return eliminatedresult eliminated

 eliminated    def eliminatedrender eliminated_text eliminated(self eliminated) eliminated-> eliminatedstr eliminated: eliminated
 eliminated        eliminatedlines eliminated= eliminatedself eliminated.render eliminated_lines eliminated() eliminated
 eliminated        eliminatedreturn eliminated"\n". eliminatedjoin eliminated(lines eliminated) eliminated
<...>
def manager() -> tuple[ConnectionManager, TopicRouter, Metrics параллеCollector паралле]: паралле
 паралле cfg паралле = паралле Config(queue_depth паралле = паралле 5)
    metrics = MetricsCollector()
<...>
    metrics = MetricsCollector()
    metrics.counter("delivery_errors").inc消费电子()消费电子()
消费电子()
    router消费电子 =消费电子 Topic消费电子 Router(cfg, metrics)
```

**Can you publish exact settings which were used to generate benchmark data? KV-cache settings, temperature and so on.**

**Comment by ABLomas** [comment]
OK, one more addition. I played with temperature a little. It seems that outputs are very sensitive to it. 0.1 is too low, 0.7 is too high, both cause loopings and shizo writing.
0.4-0.5 provide much better rate (also disabled mmproj but based on earlier tests it should not have impact on speed):
| **Task** | **Acceptance Rate** | **Prompt Speed** | **Output Speed** |
| :--- | :--- | :--- | :--- |
| **Task store** | **57.6%** | 1002.6 tok/s | **106.9 tok/s** |
| **KV report** | **53.8%** | 685.6 tok/s | **108.6 tok/s** |
| **Doubly-linked** | **47.0%** | 364.2 tok/s | **89.4 tok/s** |
| **Multi-turn** | **29.5%** | 863.9 tok/s | **55.6 tok/s** |

Code is at least looking like code, no more "budget reached" loops, no more chinese/hebrew writings. Did not tested but just by quick look it looks OK.
Much closer to provided benchmark table.
